exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/AsyncPageReader.java

```diff
@@ -284,30 +278,9 @@ protected void nextInternal() throws IOException {
 
   }
 
-  private void waitForExecutionResult() throws InterruptedException, ExecutionException {
-    // Get the execution result but don't remove the Future object from the "asyncPageRead" queue yet;
-    // this will ensure that cleanup will happen properly in case of an exception being thrown
-    asyncPageRead.peek().get(); // get the result of execution
-    // Alright now remove the Future object
-    asyncPageRead.poll();
-  }
-
   @Override public void clear() {
     //Cancelling all existing AsyncPageReaderTasks
-    while (asyncPageRead != null && !asyncPageRead.isEmpty()) {
-      try {
-        Future<Void> f = asyncPageRead.poll();
-        if(!f.isDone() && !f.isCancelled()){
-          f.cancel(true);
-        } else {
-          f.get(1, TimeUnit.MILLISECONDS);
-        }
-      } catch (RuntimeException e) {
-        // Do Nothing
-      } catch (Exception e) {
-        // Do nothing.
-      }
-    }
+    executableTasksLatch.await(() -> true);
```

Comment:
Isn't the original code doing the same thing that `await` is doing? TBH, I'd really like to understand where the memory leak was occurring. (Just trying to understand how this PR, does, in fact, fix the issue)

Reply:
No, `ExecutableTasksLatch.await()` guarantees that when it returns all tasks submitted for execution are either done or canceled. `Future.cancel()` does not wait for the `FutureTask` to be canceled as it merely interrupts the thread where `FutureTask` is running (in the case it is already running). Note that after `Future` is canceled it is not possible to check whether it is finished or not (`Future.get` throws `CancellationException`).

Missing guarantee that all tasks are finished when `clear()` returns, so some tasks continue to run and reference vectors, so allocator reports a memory leak.

Reply:
Seems to me that we could have extended FutureTask to provide this guarantee. Perhaps that is what @ilooner would have liked too.
See this : https://stackoverflow.com/questions/6040962/wait-for-cancel-on-futuretask

Reply:
I know and saw that post before the fix for the same problem was implemented only for `PartitionerDecorator`. I decided to go with a similar approach outlined in the post, but that does not use `FutureTask`to support additional functionality that I outlined in my response to @ilooner. This implementation supports asynchronous cancellation (call to `cancel()` does not block waiting for a task to complete allowing a faster cancellation) that the solution that extends `FutureTask` does not provide.